psi4/src/psi4/libfock/COSK.cc

```diff
@@ -344,16 +344,24 @@ void COSK::build_G_component(std::vector<std::shared_ptr<Matrix>>& D, std::vecto
 
     // per-thread K Matrix buffers (for accumulating thread contributions to K)
     std::vector<std::vector<SharedMatrix>> KT(njk, std::vector<SharedMatrix>(nthreads_));
+    std::vector<std::vector<SharedMatrix>> KTgrad(njk, std::vector<SharedMatrix>(nthreads_));
 
     // initialize per-thread objects
     IntegralFactory factory(primary_);
     for(size_t thread = 0; thread < nthreads_; thread++) {
         int_computers[thread] = std::shared_ptr<ElectrostaticInt>(static_cast<ElectrostaticInt *>(factory.electrostatic().release()));
         bf_computers[thread] = std::make_shared<BasisFunctions>(primary_, grid->max_points(), grid->max_functions());
+        if (do_gradient_) {
+            bf_computers[thread]->set_deriv(1);
+        }
         for(size_t jki = 0; jki < njk; jki++) {
             KT[jki][thread] = std::make_shared<Matrix>(nbf, nbf);
+            KTgrad[jki][thread] = std::make_shared<Matrix>(natom, 3);
         }
     }
+    
+    // exchange gradient
+    auto Kgrad = std::make_shared<Matrix>("Exchange Gradient", natom, 3);
```

Comment:
and here

Reply:
Moved to final if-statement block, since this is only used there.